cmd/update/src/lib.rs

```diff
@@ -0,0 +1,152 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! ## `humility update`
+//!
+//! Writes a binary to the specified update target as defined in Hubris
+//!
+//! ```console
+//! $ humility update --target ImageB update.bin
+//! humility: attached via CMSIS-DAP
+//! humility: Starting update using an update block size of 512
+//! humility: (Erase may take a moment)
+//! humility: Comitting update
+//! humility: Update done.
+//! ```
+//!
+
+use humility::cli::Subcommand;
+use humility_cmd::hiffy::*;
+use humility_cmd::{Archive, Attach, Command, Validate};
+
+use std::path::PathBuf;
+
+use anyhow::{bail, Result};
+use clap::Command as ClapCommand;
+use clap::{CommandFactory, Parser};
+use hif::*;
+
+use indicatif::{ProgressBar, ProgressStyle};
+
+extern crate log;
+
+#[derive(Parser, Debug)]
+#[clap(name = "update", about = "Write a software update")]
+struct UpdateArgs {
+    /// sets timeout
+    #[clap(
+        long, short = 'T', default_value = "50000", value_name = "timeout_ms",
+        parse(try_from_str = parse_int::parse)
+    )]
+    timeout: u32,
+
+    #[clap(long)]
+    /// Which image to update
+    target: String,
+    #[clap(long)]
+    /// Don't commit the image, only write
+    skip_commit: bool,
+    /// path to binary to write
+    path: PathBuf,
+}
+
+fn update(context: &mut humility::ExecutionContext) -> Result<()> {
+    let core = &mut **context.core.as_mut().unwrap();
+    let hubris = context.archive.as_ref().unwrap();
+    let Subcommand::Other(subargs) = context.cli.cmd.as_ref().unwrap();
+
+    let subargs = UpdateArgs::try_parse_from(subargs)?;
+    let mut context = HiffyContext::new(hubris, core, subargs.timeout)?;
+    let funcs = context.functions()?;
+
+    let start_update_fn = funcs.get("StartUpdate", 0)?;
```

Comment:
Vague architecture question: why are these specially-defined Hiffy functions (e.g. [here](https://github.com/oxidecomputer/hubris/blob/92397da6b5a7a003e63d0bc1e343b1a437002002/task/hiffy/src/stm32h7.rs#L135-L142)), instead of calls through the `Update.*` Idol API ([here](https://github.com/oxidecomputer/hubris/blob/92397da6b5a7a003e63d0bc1e343b1a437002002/idl/update.idol))?  It seems like this is an unnecessary duplication of API surface area.

Reply:
Can you give me a pointer to how you would call these through idol? This was the technique I know and love for adding humility cmmands.

Reply:
Sure, [here's how it's done in `humility auxflash`](https://github.com/oxidecomputer/humility/blob/master/cmd/auxflash/src/lib.rs#L328-L340)

This is somewhat new: it relies on `humility hiffy` support for leases, which is a recent addition!

The advantage is that – instead of needing special-case Hiffy functions and manually building HIF programs – you can call _any_ Idol interface that takes a lease, using a nice high-level API.